### ISSUE TITLE
ASP.Net 4.5 OAuth fails with a 400 response, out of the box

### ISSUE THREAD

**Issue by AArnott** [DotNetOpenAuth/DotNetOpenAuth#203]
Moved from [here](https://dotnetopenauth.uservoice.com/admin/forums/14800-general/suggestions/3156765-asp-net-4-5-oauth-fails-with-a-400-response-out-o?tracking_code=cdd20b1feeff1bc911438da0278f0680)

I fired up VS 2012 and created a new asp.net website with the oAuth templates and created an application with Facebook to test it out. I modified the AuthConfig.cs to contain the app id and app secret, but my login test is returning a 400 response, presumably from Facebook's server.

Here is the line upon which the 400 response occurs:

var authResult = OpenAuth.VerifyAuthentication(redirectUrl);

Like I said, I've modified the two lines in AuthConfig.cs, but nothing else.

"System.NullReferenceException was unhandled by user code 
HResult=-2147467261 
Message=Object reference not set to an instance of an object. 
Source=DotNetOpenAuth.AspNet"

Thoughts?

**Comment by engineeringstuff** [comment]
I have the same issue, I checked the Response property of the WebException and got this

{"error":{"message":"Error validating verification code.","type":"OAuthException","code":100}}

**Comment by engineeringstuff** [comment]
I figured it out.

Everything I read suggested that the url passed in QueryAccessToken should match the return URL given to Facebook - to me these appeared exactly the same, down to the byte level. The only difference was escaping.

In my code, my return URI makes use of Base64 encoding and therefore contains "==" in parts, the return URL that Facebook redirects to had these == escaped out with %3D.

The URL to check against in QueryAccessToken only had == escaped once (%3D%3D) but they should have been double escaped (%253D%253D).

Basically "NormalizeHexEncoding" only escapes these values once, which actually seems fine to me. The real problem to me is that Facebook is passed escaped values for == as the response URI, I don't think it should escape them at all.

To recap (because I know I've explained this kinda bad)
- 1. We auth with Facebook and say redirect to a URI with == in it
- 2. It actually redirects to a URI with %3D%3D in it
- 3. When we verify we pass it a single escaped ==
- 4. Facebook ends up comparing it's value of a URI with %3D%3D in it, compared to our given value of ==
- 5. Auth fails

I haven't dug this far yet, but it seems that between step 1 and step 2 there's a process happenning where it could be escaping the return URI too much

I verifiied this with the following (very hacky) shim in FacebookClient.cs

<pre>
                new Dictionary<string, string> {
                    { "client_id", this.appId },
                    { "redirect_uri", NormalizeHexEncoding(returnUrl.AbsoluteUri).Replace("==", "%3D%3D") },
                    { "client_secret", this.appSecret },
                    { "code", authorizationCode },
                });
</pre>

I may dig further
